File: devices-api-java-mvc/src/main/java/com/lucas/devicesapijavamvc/service/DeviceService.java
package com.lucas.devicesapijavamvc.service;

import com.lucas.devicesapijavamvc.domain.Device;
import com.lucas.devicesapijavamvc.domain.DeviceState;
import com.lucas.devicesapijavamvc.kafka.DeviceEventPublisher;
import com.lucas.devicesapijavamvc.repository.DeviceRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
public class DeviceService {

    private final DeviceRepository repo;
    private final Optional<DeviceEventPublisher> publisher;

    public DeviceService(DeviceRepository repo,
                         Optional<DeviceEventPublisher> publisher) {
        this.repo = repo;
        this.publisher = publisher;
    }

    @Transactional
    public Device create(Device newDevice) {
        Device saved = repo.save(newDevice);
        publisher.ifPresent(p -> p.publishCreated(saved, "JAVA"));
        return saved;
    }

    @Transactional(readOnly = true)
    public Device get(Long id) {
        return repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Device %d not found".formatted(id)));
    }

    @Transactional(readOnly = true)
    public List<Device> all() {
        return repo.findAll();
    }

    @Transactional(readOnly = true)
    public List<Device> byBrand(String brand) {
        return repo.findByBrand(brand);
    }

    @Transactional(readOnly = true)
    public List<Device> byState(DeviceState state) {
        return repo.findByState(state);
    }

    @Transactional
    public Device updateFull(Long id, Device incoming) {
        var existing = get(id);

        // Domain rule: name/brand cannot be changed if IN_USE
        if (existing.getState() == DeviceState.IN_USE) {
            boolean nameChanged = !existing.getName().equals(incoming.getName());
            boolean brandChanged = !existing.getBrand().equals(incoming.getBrand());
            if (nameChanged || brandChanged) {
                throw new IllegalStateException("Name and brand cannot be updated when device is in use");
            }
        }

        existing.setName(incoming.getName());
        existing.setBrand(incoming.getBrand());
        existing.setState(incoming.getState());

        // creationTime and externalId remain unchanged
        return repo.save(existing);
    }

    @Transactional
    public void delete(Long id) {
        var existing = get(id);

        // Domain rule: IN_USE devices cannot be deleted
        if (existing.getState() == DeviceState.IN_USE) {
            throw new IllegalStateException("Cannot delete device that is in use");
        }

        repo.delete(existing);
    }

    public static class NotFoundException extends RuntimeException {
        public NotFoundException(String message) {
            super(message);
        }
    }
}

File: devices-api-kotlin-reactive/src/main/kotlin/com/lucas/devicesapikotlinreactive/application/service/DeviceService.kt
package com.lucas.devicesapikotlinreactive.application.service

import com.lucas.devicesapikotlinreactive.domain.model.Device
import com.lucas.devicesapikotlinreactive.domain.model.DeviceState
import com.lucas.devicesapikotlinreactive.domain.port.DeviceEventPublisherPort
import com.lucas.devicesapikotlinreactive.domain.port.DeviceRepositoryPort
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class DeviceService(
    private val repo: DeviceRepositoryPort,
    private val eventPublisher: DeviceEventPublisherPort?
) {

    fun create(name: String, brand: String, state: DeviceState): Mono<Device> {
        val device = Device.create(name, brand, state)

        return repo.save(device)
            .flatMap { saved ->
                eventPublisher
                    ?.publishCreated(saved)
                    ?.thenReturn(saved)
                    ?: Mono.just(saved)
            }
    }

    fun get(id: Long): Mono<Device> =
        repo.findById(id)
            .switchIfEmpty(Mono.error(NotFoundException("Device $id not found")))

    fun all(): Flux<Device> = repo.findAll()

    fun byBrand(brand: String): Flux<Device> = repo.findByBrand(brand)

    fun byState(state: DeviceState): Flux<Device> = repo.findByState(state)

    fun updateFull(id: Long, incoming: Device): Mono<Device> =
        get(id).flatMap { existing ->
            // Domain rule: name/brand cannot be changed if IN_USE
            if (existing.state == DeviceState.IN_USE &&
                (existing.name != incoming.name || existing.brand != incoming.brand)
            ) {
                return@flatMap Mono.error<Device>(
                    IllegalStateException("Name and brand cannot be updated when device is in use")
                )
            }

            val updated = existing.copy(
                name = incoming.name,
                brand = incoming.brand,
                state = incoming.state
                // externalId and creationTime are preserved
            )

            repo.save(updated)
        }

    fun delete(id: Long): Mono<Void> =
        get(id).flatMap { existing ->
            // Domain rule: IN_USE devices cannot be deleted
            if (existing.state == DeviceState.IN_USE) {
                return@flatMap Mono.error<Void>(
                    IllegalStateException("Cannot delete device that is in use")
                )
            }
            repo.delete(existing.id
                ?: return@flatMap Mono.error(IllegalStateException("Cannot delete device without id"))
            )
        }

    fun upsertByExternalId(fromEvent: Device): Mono<Device> =
        repo.findByExternalId(fromEvent.externalId)
            .flatMap { existing ->
                val updated = existing.copy(
                    name = fromEvent.name,
                    brand = fromEvent.brand,
                    state = fromEvent.state
                    // externalId & creationTime preserved
                )
                repo.save(updated)
            }
            .switchIfEmpty(repo.save(fromEvent))

    class NotFoundException(message: String) : RuntimeException(message)
}
